### ISSUE TITLE
Optional Dependencies

### ISSUE THREAD

**Issue by ForbodingAngel** [beyond-all-reason/spring#605]
The ability to flag a listed dependency in modinfo as optional and not require it in order for the game to function. Of course this would only be used for non-synced things.

**Comment by sprunk** [comment]
![image](https://user-images.githubusercontent.com/2573076/215413213-4917da4e-0482-4210-82ae-e5c82240681b.png)

The request would be to add this section. If missing, loading such archive won't crash the game, and its contents would not be visible to synced VFS modes.

Normally optional stuff can just be dumped into the user's Spring folder (and read as VFS.RAW). Forb's request attempts to solve a distribution problem - his game distributes assets manually so adding this would let him tell people "place the .zip in [spring]/games/music_pack.sd7" as opposed to "unpack the zip in [spring], maintain folder hierarchy, pay attention to files that may overlap with other archives or earlier local files, ..." etc.

I think this may already be achievable without any changes though. `Spring.HasArchive` [exists](https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Lua/LuaArchive.cpp#L28) and then `Spring.UseArchive` and `Spring.MapArchive` [are also a thing](https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Lua/LuaVFS.cpp#L93-L95), I'm not sure which does what you want but presumably one of them loads an archive into VFS. So you'd do something like
```lua
if Spring.HasArchive("Music Pack v1") then -- or maybe "/games/music_v1.sd7"? idk, needs documentation with examples
  Spring.MapArchive("Music Pack v1") -- or maybe Spring.UseArchive, also needs documentation with examples
end
```

**Comment by marcushutchings** [comment]
Thank you, Sprunk, for the clarification. Oh, okay, it seems I may have mistaken what is being referred to as modinfo. Sorry. There's a file in the engine called modinfo, which reads the game archive, looks for modrules, and loads the parameters found there. I realise now that this request isn't referring that. That's not to say that it is incorrect to use the term modinfo - it may be that modinfo has been used to refer to multiple different things and I picked up on the wrong thing. So apologies for that, and thank you again, Sprunk, for the clarification.

**Comment by sprunk** [comment]
> There's a file in the engine called modinfo, which reads the game archive, looks for modrules,

Yes, game devs know that as modrules https://github.com/beyond-all-reason/spring/blob/BAR105/rts/Sim/Misc/ModInfo.cpp#L136

Modinfo being referred to here is this archive metadata file https://github.com/beyond-all-reason/spring/blob/BAR105/rts/System/FileSystem/ArchiveScanner.cpp#L98

**Comment by lhog** [comment]
@p2004a is this something doable?

**Comment by sprunk** [comment]
Put the three lines of code above in some gadget and see what happens?

**Comment by p2004a** [comment]
@ForbodingAngel I believe the suggestion by @sprunk to use the the HasArchive + Map/UseArchve is covering this feature request?

Or is there something thing else needed? If it is for the ability to easily download those optional archives then it's not function of the engine, but lobby.

**Comment by sprunk** [comment]
Lobby might still use engine-supplied unitsync library to parse the file so it might still require an engine change.

**Comment by sprunk** [comment]
Follow-up: I'm not sure if unitsync even has interfaces to get the regular dependencies. There's these https://github.com/beyond-all-reason/spring/blob/BAR105/tools/unitsync/unitsync_api.h#L185-L190 which add an archive with all its dependencies to unitsync's internal VFS but the list of dependencies itself doesn't seem to be exposed anywhere and it relies on those dependencies being distributed somehow else. I don't know what lobbies actually use to actually learn of dependencies and download them (I think ZK either parses modinfo.lua with some in-house parser or just doesn't handle dependencies? idk) but presumably optional dependencies would use whatever method is already used for regular dependencies. (Adding regular dependencies to unitsync could be a separate feature request.)

**Comment by sprunk** [comment]
I've tested and `VFS.MapArchive` works; in fact, it works _too well_ to the point it can't be exposed to unsynced as-is, see 0ee88788931f9f0b195eb5f895f1092fde4211c0
